storage-proofs/src/circuit/stacked/encoding_proof.rs

```diff
@@ -91,18 +69,38 @@ impl EncodingProof {
         )
     }
 
-    pub fn synthesize<CS: ConstraintSystem<Bls12>>(
+    pub fn synthesize_key<CS: ConstraintSystem<Bls12>>(
         self,
         mut cs: CS,
         params: &<Bls12 as JubjubEngine>::Params,
         replica_id: &[Boolean],
+        exp_encoded_node: &num::AllocatedNum<Bls12>,
     ) -> Result<(), SynthesisError> {
-        let EncodingProof {
+        let EncodingProof { node, parents } = self;
+
+        let key = Self::create_key(
+            cs.namespace(|| "create_key"),
+            params,
+            replica_id,
             node,
             parents,
-            encoded_node,
-            decoded_node,
-        } = self;
+        )?;
+
+        // enforce equality
+        constraint::equal(&mut cs, || "equality_key", &exp_encoded_node, &key);
+
+        Ok(())
+    }
+
+    pub fn synthesize_decoded<CS: ConstraintSystem<Bls12>>(
+        self,
+        mut cs: CS,
+        params: &<Bls12 as JubjubEngine>::Params,
+        replica_id: &[Boolean],
+        exp_encoded_node: &num::AllocatedNum<Bls12>,
```

Comment:
Name is confusing. `exp` sounds like it relates to `expander` in the context of SDR. I think this is now the only kind of 'encoding' that happens, but if you want to be more descriptive, I think we previously agreed that `add_encode` is what we would call this operation. (Assuming I am following.)